src/EventStore.Core/TransactionLog/Scavenging/ScavengeState.cs

```diff
@@ -0,0 +1,357 @@
+	// we don't need to store data for every original stream, only ones that need scavenging.
+	public class ScavengeState {
+		protected static readonly ILogger Log = LogManager.GetLoggerFor<ScavengeState>();
+	}
+
+	public class ScavengeState<TStreamId> : ScavengeState, IScavengeState<TStreamId> {
+		private readonly IScavengeStateBackend<TStreamId> _backend;
+		private readonly ObjectPool<IScavengeStateBackend<TStreamId>> _backendPool;
+
+		private readonly CollisionDetector<TStreamId> _collisionDetector;
+
+		// data stored keyed against metadata streams
+		private readonly MetastreamCollisionMap<TStreamId> _metastreamDatas;
+
+		// data stored keyed against original (non-metadata) streams
+		private readonly OriginalStreamCollisionMap<TStreamId> _originalStreamDatas;
+
+		private readonly IScavengeMap<int, ChunkTimeStampRange> _chunkTimeStampRanges;
+		private readonly IChunkWeightScavengeMap _chunkWeights;
+		private readonly IScavengeMap<Unit, ScavengeCheckpoint> _checkpointStorage;
+		private readonly ITransactionManager _transactionManager;
+
+		private readonly ILongHasher<TStreamId> _hasher;
+		private readonly IMetastreamLookup<TStreamId> _metastreamLookup;
+
+		public ScavengeState(
+			ILongHasher<TStreamId> hasher,
+			IMetastreamLookup<TStreamId> metastreamLookup,
+			ObjectPool<IScavengeStateBackend<TStreamId>> backendPool) {
+
+			_backendPool = backendPool;
+			_backend = _backendPool.Get();
+
+			// todo: in log v3 inject an implementation that doesn't store hash users
+			// since there are no collisions.
+			_collisionDetector = new CollisionDetector<TStreamId>(
+				hashUsers: new LruCachingScavengeMap<ulong, TStreamId>(
+					_backend.Hashes,
+					cacheMaxCount: 100_000),
+				collisionStorage: _backend.CollisionStorage,
+				hasher: hasher);
+
+			_hasher = hasher;
+			_metastreamLookup = metastreamLookup;
+			_checkpointStorage = _backend.CheckpointStorage;
+
+			_metastreamDatas = new MetastreamCollisionMap<TStreamId>(
+				_hasher,
+				_collisionDetector.IsCollision,
+				_backend.MetaStorage,
+				_backend.MetaCollisionStorage);
+
+			_originalStreamDatas = new OriginalStreamCollisionMap<TStreamId>(
+				_hasher,
+				_collisionDetector.IsCollision,
+				_backend.OriginalStorage,
+				_backend.OriginalCollisionStorage);
+
+			_chunkTimeStampRanges = _backend.ChunkTimeStampRanges;
+			_chunkWeights = _backend.ChunkWeights;
+
+			_transactionManager = _backend.TransactionManager;
+			_transactionManager.RegisterOnRollback(OnRollback);
+		}
+
+		public void Dispose() {
+			_transactionManager.UnregisterOnRollback();
+			_backendPool.Return(_backend);
+			_backendPool.Dispose();
+		}
+
+		private void OnRollback() {
+			// a transaction has been rolled back
+			// need to clear whatever we think we know about collisions to stop us mistaking a new
+			// collision for an old one.
+			_collisionDetector.ClearCaches();
+
+			// there is no need to clear the HashUsers LRU cache
```

Comment:
```suggestion
			// there is no need to clear the HashUsers LRU cache since:
			// i) a hash is added to the HashUsers map only when it is first encountered in the log. In this case, collision detection returns: NoCollision
			// ii) if a hash is already in the HashUsers map when it is first encountered in the log (i.e after a rollback), the stream ids in the map and log will match and collision detection also returns: NoCollision
			// Thus, the net effect is the same after processing the newly encountered hash in cases i) and ii).
```

Reply:
somewhere along the line i removed this comment anyway

Reply:
I suggest we keep at least the original comment just to make sure people who look at the code in the future don't think that it was an oversight but was by design.

Reply:
ok cool :)